pallets/liquidity-pools-gateway-queue/src/lib.rs

```diff
@@ -200,17 +208,34 @@ pub mod pallet {
 		}
 
 		fn service_message_queue(max_weight: Weight) -> Weight {
-			let mut weight_used = Weight::zero();
+			let mut last_processed_nonce = LastProcessedNonce::<T>::get();
+
+			// 1 read for the last processed nonce
+			let mut weight_used = T::DbWeight::get().reads(1);
+
+			loop {
+				if let Err(_) = last_processed_nonce.ensure_add_assign(One::one()) {
+					Self::deposit_event(Event::<T>::MaxNumberOfMessagesWasReached);
+
+					break;
+				}
 
-			let mut nonces = MessageQueue::<T>::iter_keys().collect::<Vec<_>>();
-			nonces.sort();
+				// 1 read for the nonce
+				weight_used.saturating_accrue(T::DbWeight::get().reads(1));
 
-			for nonce in nonces {
-				let message =
-					MessageQueue::<T>::get(nonce).expect("valid nonce ensured by `iter_keys`");
+				if last_processed_nonce > MessageNonceStore::<T>::get() {
+					break;
+				}
 
+				// 1 read for the message
 				weight_used.saturating_accrue(T::DbWeight::get().reads(1));
 
+				let message = match MessageQueue::<T>::get(last_processed_nonce) {
+					Some(msg) => msg,
+					// No message found, we can stop.
+					None => break,
+				};
```

Comment:
I think the idea is fine! But here we could have an undesired behavior:

A message could be processed with priority through `process_message()` and consequently be extracted from the `MessageQueue`. If that happens, then `last_processed_nonce` could not be found, but that doesn't mean that we want to end up iterating the message queue. TLDR: Could be holes in the `MessageQueue` storage.

We should somehow jump over the holes and continue iterating until reaching the maximum nonce added to the `MessageQueue`

Reply:
Maybe another storage value called `LastKnownNonce` or similar does the trick

Reply:
`MessageNonceStore` contains the last inserted nonce, I will adapt this as per your comment, I forgot about `process_message` when tackling this. Thank you!

Reply:
This is one of the whys why having that method increase complexity 😅